SmallRye GraphQL conflict with same enum names in different models

**SmallRye GraphQL cannot differentiate same enum names in different models from version 1.6.0 to 1.8.1.**
We have GraphQL API for banking. We have separate models i.e `Accounts` and `Balances`. In `Accounts`  we have an enum called `Type` and we have enum with same name `Type` in `Balances`.

**The expected behavior is to see they're working fine as in previous versions.**
The quarkus version was 1.5.2 for this apps and we decide to update the quarkus and plug-in versions. In the all versions after 1.5.2, the problem is same. In 1.5.2, it perfectly works without having conflict even if we have same enum name in different models.

**Actual behavior**
 The data is stored in Neo4j and when we run the application, it works fine with accounts. But when running the `Balances`, it gives the following stacktrace:
```
2020-09-29 18:51:46,311 ERROR [io.qua.run.boo.StartupActionImpl] (Quarkus Main Thread) Error running Quarkus: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.runner.bootstrap.StartupActionImpl$3.run(StartupActionImpl.java:134)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
        at java.base/java.lang.Class.newInstance(Class.java:584)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:60)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:106)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
        ... 6 more
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:287)
        ... 15 more
Caused by: graphql.AssertException: type can't be null
        at graphql.Assert.assertNotNull(Assert.java:16)
        at graphql.schema.GraphQLFieldDefinition.<init>(GraphQLFieldDefinition.java:85)
        at graphql.schema.GraphQLFieldDefinition$Builder.build(GraphQLFieldDefinition.java:473)
        at io.smallrye.graphql.bootstrap.Bootstrap.createGraphQLFieldDefinitionFromField(Bootstrap.java:399)
        at io.smallrye.graphql.bootstrap.Bootstrap.createGraphQLFieldDefinitionsFromFields(Bootstrap.java:386)
        at io.smallrye.graphql.bootstrap.Bootstrap.createGraphQLObjectType(Bootstrap.java:284)
        at io.smallrye.graphql.bootstrap.Bootstrap.createGraphQLObjectTypes(Bootstrap.java:271)
        at io.smallrye.graphql.bootstrap.Bootstrap.generateGraphQLSchema(Bootstrap.java:123)
        at io.smallrye.graphql.bootstrap.Bootstrap.bootstrap(Bootstrap.java:103)
        at io.smallrye.graphql.cdi.producer.GraphQLProducer.initialize(GraphQLProducer.java:41)
        at io.smallrye.graphql.cdi.producer.GraphQLProducer.initialize(GraphQLProducer.java:37)
        at io.smallrye.graphql.cdi.producer.GraphQLProducer_ClientProxy.initialize(GraphQLProducer_ClientProxy.zig:185)
        at io.quarkus.smallrye.graphql.runtime.SmallRyeGraphQLRecorder.createExecutionService(SmallRyeGraphQLRecorder.java:24)
        at io.quarkus.deployment.steps.SmallRyeGraphQLProcessor$buildExecutionService-1556698146.deploy_2(SmallRyeGraphQLProcessor$buildExecutionService-1556698146.zig:5278)
        at io.quarkus.deployment.steps.SmallRyeGraphQLProcessor$buildExecutionService-1556698146.deploy(SmallRyeGraphQLProcessor$buildExecutionService-1556698146.zig:50)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:267)
```



**To Reproduce**
Steps to reproduce the behavior:
1. Create 2 separate models.
2. Have an enum type in both models with same enum name.(i.e enum `Type` in `Accounts` and `Balances`)
3. Run it in debug mode, put the breakpoint on line 85 in `GraphQLFieldDefinition.java` and see the hitting `type can't be null` on same enum name.



**Environment (please complete the following information):**
 - MSYS_NT-10.0-19041 DESKTOP-6RLSJFU 3.0.7-338.x86_64 2019-11-21 23:07 UTC x86_64 Msys
 - Output of `java -version`: openjdk version "11.0.7" 2020-04-14
 - GraalVM version (if different from Java): OpenJDK Runtime Environment GraalVM CE 20.1.0 (build 11.0.7+10-jvmci-20.1-b02)
 - Quarkus version or git rev: 1.8.1
 - Apache Maven 3.6.3

**Additional context**
If you really want to see that's the reason, you can simply create other models as well and have same enum name. It will hit on that and will say `type can't be null`

Regards!
Yusuf

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -38,7 +38,7 @@
         <smallrye-health.version>2.2.3</smallrye-health.version>
         <smallrye-metrics.version>2.4.4</smallrye-metrics.version>
         <smallrye-open-api.version>2.0.10</smallrye-open-api.version>
-        <smallrye-graphql.version>1.0.11</smallrye-graphql.version>
+        <smallrye-graphql.version>1.0.12</smallrye-graphql.version>
         <smallrye-opentracing.version>1.3.4</smallrye-opentracing.version>
         <smallrye-fault-tolerance.version>4.3.1</smallrye-fault-tolerance.version>
         <smallrye-jwt.version>2.3.0</smallrye-jwt.version>

--- a/extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLConfig.java
+++ b/extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLConfig.java
@@ -3,6 +3,7 @@ package io.quarkus.smallrye.graphql.deployment;
 import io.quarkus.runtime.annotations.ConfigDocSection;
 import io.quarkus.runtime.annotations.ConfigItem;
 import io.quarkus.runtime.annotations.ConfigRoot;
+import io.smallrye.graphql.schema.helper.TypeAutoNameStrategy;
 
 @ConfigRoot(name = "smallrye-graphql")
 public class SmallRyeGraphQLConfig {
@@ -18,6 +19,12 @@ public class SmallRyeGraphQLConfig {
      */
     @ConfigItem(name = "metrics.enabled", defaultValue = "false")
     boolean metricsEnabled;
+
+    /**
+     * Change the type naming strategy.
+     */
+    @ConfigItem(defaultValue = "Default")
+    TypeAutoNameStrategy autoNameStrategy;
 
     /**
      * UI configuration

--- a/extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java
+++ b/extensions/smallrye-graphql/deployment/src/main/java/io/quarkus/smallrye/graphql/deployment/SmallRyeGraphQLProcessor.java
@@ -69,6 +69,7 @@ import io.smallrye.graphql.schema.Annotations;
 import io.smallrye.graphql.schema.SchemaBuilder;
 import io.smallrye.graphql.schema.model.Argument;
 import io.smallrye.graphql.schema.model.Field;
+import io.smallrye.graphql.schema.model.Group;
 import io.smallrye.graphql.schema.model.InputType;
 import io.smallrye.graphql.schema.model.InterfaceType;
 import io.smallrye.graphql.schema.model.Operation;
@@ -155,7 +156,8 @@ public class SmallRyeGraphQLProcessor {
             CombinedIndexBuildItem combinedIndex) {
 
         IndexView index = combinedIndex.getIndex();
-        Schema schema = SchemaBuilder.build(index);
+
+        Schema schema = SchemaBuilder.build(index, quarkusConfig.autoNameStrategy);
 
         recorder.createExecutionService(beanContainer.getValue(), schema);
 
@@ -251,7 +253,9 @@ public class SmallRyeGraphQLProcessor {
         Set<String> classes = new HashSet<>();
 
         classes.addAll(getOperationClassNames(schema.getQueries()));
+        classes.addAll(getOperationClassNames(schema.getGroupedQueries()));
         classes.addAll(getOperationClassNames(schema.getMutations()));
+        classes.addAll(getOperationClassNames(schema.getGroupedMutations()));
         classes.addAll(getTypeClassNames(schema.getTypes().values()));
         classes.addAll(getInputClassNames(schema.getInputs().values()));
         classes.addAll(getInterfaceClassNames(schema.getInterfaces().values()));
@@ -289,6 +293,15 @@ public class SmallRyeGraphQLProcessor {
                 classes.addAll(getAllReferenceClasses(argument.getReference()));
             }
             classes.addAll(getAllReferenceClasses(operation.getReference()));
+        }
+        return classes;
+    }
+
+    private Set<String> getOperationClassNames(Map<Group, Set<Operation>> groupedOperations) {
+        Set<String> classes = new HashSet<>();
+        Collection<Set<Operation>> operations = groupedOperations.values();
+        for (Set<Operation> operationSet : operations) {
+            classes.addAll(getOperationClassNames(operationSet));
         }
         return classes;
     }

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/AbstractGraphQLTest.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/AbstractGraphQLTest.java
@@ -2,6 +2,7 @@ package io.quarkus.smallrye.graphql.deployment;
 
 import java.io.IOException;
 import java.io.StringWriter;
+import java.util.Map;
 import java.util.Properties;
 
 import javax.json.Json;
@@ -61,8 +62,15 @@ public abstract class AbstractGraphQLTest {
     }
 
     protected static String getPropertyAsString() {
+        return getPropertyAsString(null);
+    }
+
+    protected static String getPropertyAsString(Map<String, String> otherProperties) {
         try {
             StringWriter writer = new StringWriter();
+            if (otherProperties != null) {
+                PROPERTIES.putAll(otherProperties);
+            }
             PROPERTIES.store(writer, "Test Properties");
             return writer.toString();
         } catch (IOException ex) {

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/Book.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/Book.java
@@ -1,0 +1,23 @@
+package io.quarkus.smallrye.graphql.deployment;
+
+import java.time.LocalDate;
+import java.util.Arrays;
+import java.util.List;
+
+public class Book {
+    public String isbn;
+    public String title;
+    public LocalDate published;
+    public List<String> authors;
+
+    public Book() {
+    }
+
+    public Book(String isbn, String title, LocalDate published, String... authors) {
+        this.isbn = isbn;
+        this.title = title;
+        this.published = published;
+        this.authors = Arrays.asList(authors);
+    }
+
+}

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/BookGraphQLApi.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/BookGraphQLApi.java
@@ -1,0 +1,45 @@
+package io.quarkus.smallrye.graphql.deployment;
+
+import java.time.LocalDate;
+import java.time.Month;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.microprofile.graphql.Description;
+import org.eclipse.microprofile.graphql.GraphQLApi;
+import org.eclipse.microprofile.graphql.Mutation;
+import org.eclipse.microprofile.graphql.Name;
+import org.eclipse.microprofile.graphql.Query;
+
+@GraphQLApi
+@Name("books")
+@Description("Allow all book releated APIs")
+public class BookGraphQLApi {
+
+    @Query
+    public List<Book> getAllBooks() {
+        return new ArrayList<>(BOOKS.values());
+    }
+
+    @Query
+    public Book getBook(String name) {
+        return BOOKS.get(name);
+    }
+
+    @Mutation
+    public Book addBook(Book book) {
+        BOOKS.put(book.title, book);
+        return book;
+    }
+
+    private static Map<String, Book> BOOKS = new HashMap<>();
+    static {
+        Book book1 = new Book("0-571-05686-5", "Lord of the Flies", LocalDate.of(1954, Month.SEPTEMBER, 17), "William Golding");
+        BOOKS.put(book1.title, book1);
+
+        Book book2 = new Book("0-582-53008-3", "Animal Farm", LocalDate.of(1945, Month.AUGUST, 17), "George Orwell");
+        BOOKS.put(book2.title, book2);
+    }
+}

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLGroupedTest.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLGroupedTest.java
@@ -1,0 +1,56 @@
+package io.quarkus.smallrye.graphql.deployment;
+
+import static io.quarkus.smallrye.graphql.deployment.AbstractGraphQLTest.MEDIATYPE_JSON;
+
+import org.hamcrest.CoreMatchers;
+import org.jboss.logging.Logger;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+/**
+ * Basic tests for Grouping.
+ */
+public class GraphQLGroupedTest extends AbstractGraphQLTest {
+
+    private static final Logger LOG = Logger.getLogger(GraphQLGroupedTest.class);
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(BookGraphQLApi.class, Book.class)
+                    .addAsResource(new StringAsset(getPropertyAsString()), "application.properties")
+                    .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml"));
+
+    @Test
+    public void testSourcePost() {
+        String booksRequest = getPayload("{\n" +
+                "  books{\n" +
+                "    book(name:\"Lord of the Flies\"){\n" +
+                "      title\n" +
+                "      authors\n" +
+                "      published\n" +
+                "    }\n" +
+                "  }\n" +
+                "}");
+
+        RestAssured.given().when()
+                .accept(MEDIATYPE_JSON)
+                .contentType(MEDIATYPE_JSON)
+                .body(booksRequest)
+                .post("/graphql")
+                .then()
+                .assertThat()
+                .statusCode(200)
+                .and()
+                .body(CoreMatchers.containsString("Lord of the Flies"));
+
+    }
+
+}

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLNamingTest.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLNamingTest.java
@@ -1,0 +1,60 @@
+package io.quarkus.smallrye.graphql.deployment;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.jboss.logging.Logger;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+import io.restassured.response.Response;
+import io.restassured.specification.RequestSpecification;
+
+/**
+ * Basic tests. POST
+ */
+public class GraphQLNamingTest extends AbstractGraphQLTest {
+
+    private static final Logger LOG = Logger.getLogger(GraphQLNamingTest.class);
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(TestResource.class, TestPojo.class, TestRandom.class, TestGenericsPojo.class)
+                    .addAsResource(new StringAsset(getPropertyAsString(configuration())), "application.properties")
+                    .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml"));
+
+    @Test
+    public void testSchema() {
+        RequestSpecification request = RestAssured.given();
+        request.accept(MEDIATYPE_TEXT);
+        request.contentType(MEDIATYPE_TEXT);
+        Response response = request.get("/graphql/schema.graphql");
+        String body = response.body().asString();
+        LOG.error(body);
+
+        Assertions.assertEquals(200, response.statusCode());
+        Assertions.assertTrue(body.contains("\"Query root\""));
+        Assertions.assertTrue(body.contains("type Query {"));
+        Assertions.assertTrue(body.contains("ping: TestPojo"));
+        Assertions.assertTrue(body.contains("generics: TestGenericsPojo_String"));
+        Assertions.assertTrue(body.contains("type TestGenericsPojo_String {"));
+        Assertions.assertTrue(body.contains("enum SomeEnum {"));
+        Assertions.assertTrue(body.contains("enum TestPojoNumber {")); // This is the important part. TestPojo merged with Number
+    }
+
+    private static Map<String, String> configuration() {
+        Map<String, String> m = new HashMap<>();
+        m.put("quarkus.smallrye-graphql.auto-name-strategy", "MergeInnerClass");
+        return m;
+
+    }
+
+}

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLTest.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/GraphQLTest.java
@@ -47,6 +47,7 @@ public class GraphQLTest extends AbstractGraphQLTest {
         Assertions.assertTrue(body.contains("generics: TestGenericsPojo_String"));
         Assertions.assertTrue(body.contains("type TestGenericsPojo_String {"));
         Assertions.assertTrue(body.contains("enum SomeEnum {"));
+        Assertions.assertTrue(body.contains("enum Number {"));
     }
 
     @Test

--- a/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/TestPojo.java
+++ b/extensions/smallrye-graphql/deployment/src/test/java/io/quarkus/smallrye/graphql/deployment/TestPojo.java
@@ -10,6 +10,8 @@ public class TestPojo {
     private String message;
     private List<String> list = Arrays.asList("a", "b", "c");
 
+    private Number number;
+
     public TestPojo() {
         super();
     }
@@ -35,9 +37,22 @@ public class TestPojo {
         this.list = list;
     }
 
-    @Override
-    public String toString() {
-        return "TestPojo{" + "message=" + message + ", list=" + list + '}';
+    public Number getNumber() {
+        return number;
     }
 
+    public void setNumber(Number number) {
+        this.number = number;
+    }
+
+    @Override
+    public String toString() {
+        return "TestPojo{" + "message=" + message + ", list=" + list + ", number=" + number + '}';
+    }
+
+    enum Number {
+        ONE,
+        TWO,
+        THREE
+    }
 }

--- a/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java
+++ b/extensions/smallrye-graphql/runtime/src/main/java/io/quarkus/smallrye/graphql/runtime/SmallRyeGraphQLRecorder.java
@@ -9,6 +9,7 @@ import io.quarkus.runtime.annotations.Recorder;
 import io.quarkus.security.identity.CurrentIdentityAssociation;
 import io.quarkus.smallrye.graphql.runtime.spi.QuarkusClassloadingService;
 import io.quarkus.vertx.http.runtime.CurrentVertxRequest;
+import io.smallrye.graphql.cdi.config.GraphQLConfig;
 import io.smallrye.graphql.cdi.producer.GraphQLProducer;
 import io.smallrye.graphql.schema.model.Schema;
 import io.vertx.core.Handler;
@@ -21,7 +22,8 @@ public class SmallRyeGraphQLRecorder {
 
     public void createExecutionService(BeanContainer beanContainer, Schema schema) {
         GraphQLProducer graphQLProducer = beanContainer.instance(GraphQLProducer.class);
-        graphQLProducer.initialize(schema);
+        GraphQLConfig graphQLConfig = beanContainer.instance(GraphQLConfig.class);
+        graphQLProducer.initialize(schema, graphQLConfig);
     }
 
     public Handler<RoutingContext> executionHandler(boolean allowGet) {

